model-optimizer/extensions/front/onnx/loop_ext.py

```diff
@@ -0,0 +1,209 @@
+    assert internal_result.soft_get('op') == 'Result'
+    assert internal_result.id in loop_node.body
+
+    loop_node.output_port_map.append({'axis': axis, 'stride': stride, 'part_size': part_size, 'start': start,
+                                      'end': end, 'external_port_id': loop_output_port_idx,
+                                      'internal_layer_id': internal_result['internal_layer_id']})
+
+
+def connect_body_input(loop_node: Node, loop_input_port_idx: int, body_parameter: Node,
+                       axis: [int, None] = None, start: [int, None] = None, end: [int, None] = None,
+                       stride: [int, None] = None, part_size: [int, None] = None):
+    assert loop_node.soft_get('op') == 'Loop'
+    assert body_parameter.soft_get('op') == 'Parameter'
+    assert body_parameter.id in loop_node.body
+
+    loop_node.input_port_map.append({'axis': axis, 'stride': stride, 'part_size': part_size, 'start': start,
+                                     'end': end, 'external_port_id': loop_input_port_idx,
+                                     'internal_layer_id': body_parameter['internal_layer_id']})
+
+
+class LoopExtractor(FrontExtractorOp):
+    op = 'Loop'
+    enabled = True
+
+    @classmethod
+    def extract(cls, loop_node):
+        Loop.update_node_stat(loop_node, {})
+
+        body_graph_proto = onnx_attr(loop_node, 'body', 'g', None)
+        main_graph = loop_node.graph
+
+        # create a Graph object for the body and take graph attributes from the main graph
+        body_graph = Graph()
+        main_graph_attrs_copy = copy.deepcopy(main_graph.graph)
+        del main_graph_attrs_copy['tensor_mapping']
+        body_graph.graph.update(main_graph_attrs_copy)
+        loop_node['body'] = body_graph
+
+        # maps a tensor name to a node produced it and the node port: str -> (node_id, node_port)
+        data_nodes_map = {}
+        body_graph.graph['tensor_mapping'] = data_nodes_map  # save mapping for possible Loop inside the Loop
+
+        body_parameters = add_initializers_and_inputs_to_graph(body_graph, body_graph_proto, data_nodes_map)
+
+        external_edges = []  # (src_node, src_out_port), dest_body_parameter_node
+        additional_params = {}  # (src_node, src_out_port) -> parameter_node (for manually added Parameters)
+        # Go through all nodes in the original model order because data nodes are defined on-the-fly and order matters
+        for pb_node in body_graph_proto.node:
+            # create an NX node
+            id = body_graph.unique_id(node_id(pb_node))
+            body_graph.add_node(id, pb=pb_node, kind='op')
+
+            # add incoming edges based on data_nodes_map
+            for dst_port, inp in enumerate(pb_node.input):
+                # should add edge inp --> id
+                if inp not in data_nodes_map:
+                    if inp == '':
+                        # input is omitted; most likely it corresponds to an optional input for an operator
```

Comment:
I guess we need to add some logic here to make Const(true) -> conditional output subgraph in the body

Reply:
Will be a part of another PR.